src/ioctls/vm.rs

```diff
@@ -1261,62 +1261,94 @@ mod tests {
-            assert!(vm_fd.create_irq_chip().is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd1, 4).is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd2, 8).is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd3, 4).is_ok());
-        }
 
-        // On aarch64, this fails because setting up the interrupt controller is mandatory before
-        // registering any IRQ.
+        assert!(vm_fd.create_irq_chip().is_ok());
+
+        assert!(vm_fd.register_irqfd(&evtfd1, 4).is_ok());
+        assert!(vm_fd.register_irqfd(&evtfd2, 8).is_ok());
+        assert!(vm_fd.register_irqfd(&evtfd3, 4).is_ok());
+        assert!(vm_fd.unregister_irqfd(&evtfd2, 8).is_ok());
+        // KVM irqfd doesn't report failure on this case:(
+        assert!(vm_fd.unregister_irqfd(&evtfd2, 8).is_ok());
+
+        // Duplicated eventfd registration.
         // On x86_64 this fails as the event fd was already matched with a GSI.
         assert!(vm_fd.register_irqfd(&evtfd3, 4).is_err());
         assert!(vm_fd.register_irqfd(&evtfd3, 5).is_err());
+        // KVM irqfd doesn't report failure on this case:(
+        assert!(vm_fd.unregister_irqfd(&evtfd3, 5).is_ok());
     }
 
     #[test]
-    #[cfg(any(
-        target_arch = "x86",
-        target_arch = "x86_64",
-        target_arch = "arm",
-        target_arch = "aarch64"
-    ))]
-    fn test_unregister_irqfd() {
+    #[cfg(target_arch = "aarch64")]
+    fn test_register_unregister_irqfd() {
         let kvm = Kvm::new().unwrap();
         let vm_fd = kvm.create_vm().unwrap();
         let evtfd1 = EventFd::new(EFD_NONBLOCK).unwrap();
         let evtfd2 = EventFd::new(EFD_NONBLOCK).unwrap();
         let evtfd3 = EventFd::new(EFD_NONBLOCK).unwrap();
-        if cfg!(any(target_arch = "x86", target_arch = "x86_64")) {
-            assert!(vm_fd.create_irq_chip().is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd1, 4).is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd2, 8).is_ok());
-            assert!(vm_fd.register_irqfd(&evtfd3, 4).is_ok());
-            assert!(vm_fd.unregister_irqfd(&evtfd2, 8).is_ok());
-            // KVM irqfd doesn't report failure on this case:(
-            assert!(vm_fd.unregister_irqfd(&evtfd2, 8).is_ok());
-        }
+
+        let mut gic_device = kvm_bindings::kvm_create_device {
+            type_: kvm_device_type_KVM_DEV_TYPE_ARM_VGIC_V3,
+            fd: 0,
+            flags: 0,
+        };
+
+        let vgic_fd = vm_fd.create_device(&mut gic_device);
+        let vgic_fd = match vgic_fd {
+            Ok(fd) => fd,
+            Err(_) => {
+                gic_device.type_ = kvm_device_type_KVM_DEV_TYPE_ARM_VGIC_V2;
+                vm_fd
+                    .create_device(&mut gic_device)
+                    .expect("Cannot create KVM vGIC device")
+            }
+        };
+
+        let raw_fd = unsafe { libc::dup(vgic_fd.as_raw_fd()) };
+        assert!(raw_fd >= 0);
+        let vgic_fd = unsafe { DeviceFd::from_raw_fd(raw_fd) };
```

Comment:
Why is this necessary? `vgic_fd` is already a `DeviceFd` object.

Why `dup` its underlying `fd`, close the original `fd` (through dropping the original `vgic_fd`), then creating another `DeviceFd` from the duplicated `fd`?

The test should work the same without these 3 lines, right?

Production code is not playing with the underlying `fd`s like this, so unless you want to exercise some path that I'm missing, just use the original `DeviceFd` provided by `vm_fd.create_device()`.

Reply:
Ah yes of course, this is just because I saw similar code in [device.rs#97](https://github.com/rust-vmm/kvm-ioctls/blob/5a50456cd1afa18e079c97bf2e87be2572b4a3f9/src/ioctls/device.rs#L97) and was not sure whether I should adding them.....The test works for both cases.

Reply:
Then I just delete them :) Thanks for pointing it out!